sw/device/lib/dif/dif_gpio.h

```diff
@@ -22,357 +22,401 @@ extern "C" {
 #endif  // __cplusplus
 
 /**
- * Configuration for initializing a GPIO device.
+ * A toggle state: enabled, or disabled.
+ *
+ * This enum may be used instead of a `bool` when describing an enabled/disabled
+ * state.
+ *
+ * This enum may be used with `dif_gpio_toggle_vec_t` to set individual bits
+ * within it; `dif_gpio_toggle_t`'s variants are guaranteed to be compatible
+ * with `dif_gpio_toggle_vec_t`.
+ */
+typedef enum dif_gpio_toggle {
+  /*
+   * The "enabled" state.
+   */
+  kDifGpioToggleEnabled = true,
+  /**
+   * The "disabled" state.
+   */
+  kDifGpioToggleDisabled = false,
+} dif_gpio_toggle_t;
+
+/**
+ * Hardware instantiation parameters for GPIO.
+ *
+ * This struct describes information about the underlying hardware that is
+ * not determined until the hardware design is used as part of a top-level
+ * design.
  */
-typedef struct dif_gpio_config { mmio_region_t base_addr; } dif_gpio_config_t;
+typedef struct dif_gpio_params {
+  /**
+   * The base address for the GPIO hardware registers.
+   */
+  mmio_region_t base_addr;
+} dif_gpio_params_t;
 
 /**
- * Internal state of a GPIO device.
+ * A handle to GPIO.
  *
- * Instances of this struct must be initialized by `dif_gpio_init()` before
- * being passed to other functions.
+ * This type should be treated as opaque by users.
  */
-typedef struct dif_gpio { mmio_region_t base_addr; } dif_gpio_t;
+typedef struct dif_gpio { dif_gpio_params_t params; } dif_gpio_t;
 
 /**
- * Generic return codes for the functions in this library.
+ * The result of a GPIO operation.
  */
 typedef enum dif_gpio_result {
   /**
-   * The function succeeded.
+   * Indicates that the operation succeeded.
    */
-  kDifGpioOk,
+  kDifGpioOk = 0,
   /**
-   * The function failed a non-specific assertion.
-   *
-   * This error is not recoverable.
+   * Indicates some unspecified failure.
    */
-  kDifGpioError,
+  kDifGpioError = 1,
   /**
-   * There is a problem with the argument(s) to the function.
+   * Indicates that some parameter passed into a function failed a
+   * precondition.
    *
-   * This return code is only returned before the function has any side effects.
-   *
-   * This error is recoverable and the operation can be retried after correcting
-   * the problem with the argument(s).
+   * When this value is returned, no hardware operations occured.
    */
-  kDifGpioBadArg,
-
+  kDifGpioBadArg = 2,
 } dif_gpio_result_t;
 
 /**
- * Set of allowed interrupt trigger configurations.
+ * A GPIO interrupt request trigger.
+ *
+ * Each GPIO pin has an associated interrupt; this enum describes which signal
+ * states on a given pin will fire its associated interrupt.
  */
-typedef enum dif_gpio_irq {
+typedef enum dif_gpio_irq_trigger {
```

Comment:
Names of the enum members should be updated. (The suggestion below already includes this but requires other usages to be updated).